packages/react/use-controllable-state/src/use-controllable-state.tsx

```diff
@@ -1,57 +1,96 @@
-}: UseControllableStateParams<T>) {
-  const [uncontrolledProp, setUncontrolledProp] = useUncontrolledState({ defaultProp, onChange });
+  caller,
+}: UseControllableStateParams<T>): [T, SetStateFn<T>] {
+  const [uncontrolledProp, setUncontrolledProp, onChangeRef] = useUncontrolledState({
+    defaultProp,
+    onChange,
+  });
   const isControlled = prop !== undefined;
   const value = isControlled ? prop : uncontrolledProp;
-  const handleChange = useCallbackRef(onChange);
 
-  const setValue: React.Dispatch<React.SetStateAction<T | undefined>> = React.useCallback(
+  // OK to disable conditionally calling hooks here because they will always run
+  // consistently in the same environment. Bundlers should be able to remove the
+  // code block entirely in production.
+  /* eslint-disable react-hooks/rules-of-hooks */
+  if (process.env.NODE_ENV !== 'production') {
+    const isControlledRef = React.useRef(prop !== undefined);
+    React.useEffect(() => {
+      const wasControlled = isControlledRef.current;
+      if (wasControlled !== isControlled) {
+        const from = wasControlled ? 'controlled' : 'uncontrolled';
+        const to = isControlled ? 'controlled' : 'uncontrolled';
+        console.warn(
+          `${caller} is changing from ${from} to ${to}. Components should not switch from controlled to uncontrolled (or vice versa). Decide between using a controlled or uncontrolled value for the lifetime of the component.`
+        );
+      }
+      isControlledRef.current = isControlled;
+    }, [isControlled, caller]);
+  }
+  /* eslint-enable react-hooks/rules-of-hooks */
+
+  const setValue = React.useCallback<SetStateFn<T>>(
     (nextValue) => {
       if (isControlled) {
-        const setter = nextValue as SetStateFn<T>;
-        const value = typeof nextValue === 'function' ? setter(prop) : nextValue;
-        if (value !== prop) handleChange(value as T);
+        const value = isFunction(nextValue) ? nextValue(prop) : nextValue;
+        if (value !== prop) {
+          onChangeRef.current?.(value);
+        }
       } else {
         setUncontrolledProp(nextValue);
       }
     },
-    [isControlled, prop, setUncontrolledProp, handleChange]
+    [isControlled, prop, setUncontrolledProp, onChangeRef]
   );
 
-  return [value, setValue] as const;
+  return [value, setValue];
 }
 
 function useUncontrolledState<T>({
   defaultProp,
   onChange,
-}: Omit<UseControllableStateParams<T>, 'prop'>) {
-  const uncontrolledState = React.useState<T | undefined>(defaultProp);
-  const [value] = uncontrolledState;
+}: Omit<UseControllableStateParams<T>, 'prop'>): [
+  Value: T,
+  setValue: React.Dispatch<React.SetStateAction<T>>,
+  OnChangeRef: React.RefObject<ChangeHandler<T> | undefined>,
+] {
+  const [value, setValue] = React.useState(defaultProp);
   const prevValueRef = React.useRef(value);
-  const handleChange = useCallbackRef(onChange);
+
+  const onChangeRef = React.useRef(onChange);
+  useInsertionEffect(() => {
```

Comment:
Using `useInsertionEffect` (or pre-19: `useLayoutEffect`) ensures we save the callback as early as possible. There may be some edge cases where the reference is updated outside of the React lifecycle for some reason, but that would be the case with our current implementation as well.